### DISCUSSION TITLE
Modules auto depend on?

### DISCUSSION THREAD

**Discussion by wakuflair** [abpframework/abp#14841] [Ideas]
Current ABP's module system needs explicitly adding `DependsOn` attributes on a module class to declare its module dependencies.

IMO if a module A referenced another module B, A must depend on B right?
So is `DependsOn` necessary actually?

Let's think it in the opposite way, is there a case that module A has referenced module B, but A does not `DependsOn` B?

If there is no such case, is it possible to implement something like "Auto Depends On"?
That is, once you reference another ABP module in your module, you automatically create an implicitly dependency relationship with it.

**Comment by EngincanV** [comment]
I don't think this would be performant since there are too many things that need to be done. First, we need to get the package name from the `*.csproj` file, then load the assembly and do the same thing for it (because a module depends on another module most likely) and find the module class for each module and register its services and middlewares and so on (too many other things that did not come to my mind now)... And you can imagine the sub-module registrations and all the other stuff. It would be cumbersome and it's a hard thing to do.

It's much easier to manually add **DependsOn** and you have more control over it.

**Comment by wakuflair** [comment]
Hi @EngincanV , thanks for your reply.

From my understanding, "many things" you mentioned has already been done in [Plugin](https://docs.abp.io/en/abp/latest/PlugIn-Modules) system.
I have not deeply looked at Plugin's source, but I think the general way of doing this is :
1. scan the dlls
2. load assemblies via reflection
3. find module classes
4. register them

"Auto Depend On" doesn't need first 2 steps since it still needs to reference packages manually, step 3 and 4 are same as Plugin.
Maybe I missed something, but I think it's pretty easy.

The key thing is that regardless of the implementation efforts, is "Auto Depend On" reasonable ?

**Reply by EngincanV** [reply]
I think it's pretty reasonable and you're right by using the Plug-in approach it can be implemented more easily. The only drawback is it would be less performant than the current system.
